File: packages/react-native-nitro-update/android/src/main/java/com/margelo/nitro/nitroupdate/OtaStorage.kt
package com.margelo.nitro.nitroupdate

import android.content.Context
import android.content.SharedPreferences
import com.margelo.nitro.NitroModules
import org.json.JSONArray
import java.io.File

internal object OtaStorage {
  /** Set by BackgroundUpdateWorker so storage uses app context when React isn't ready. */
  @Volatile
  var overrideContext: Context? = null

  private const val PREFS_NAME = "nitroupdate"
  private const val KEY_VERSION = "version"
  private const val KEY_BUNDLE_PATH = "bundlePath"
  private const val KEY_PENDING_VALIDATION = "pendingValidation"
  private const val KEY_BLACKLIST = "blacklist"
  private const val KEY_ROLLBACK_HISTORY = "rollbackHistory"
  private const val KEY_PREVIOUS_VERSION = "previousVersion"
  private const val KEY_PREVIOUS_BUNDLE_PATH = "previousBundlePath"
  private const val KEY_LAST_CHECKED_REMOTE_VERSION = "lastCheckedRemoteVersion"
  private const val KEY_BG_VERSION_CHECK_URL = "bgVersionCheckUrl"
  private const val KEY_BG_DOWNLOAD_URL = "bgDownloadUrl"
  private const val KEY_BG_BUNDLE_PATH_IN_ZIP = "bgBundlePathInZip"
  private const val KEY_BG_INTERVAL_SECONDS = "bgIntervalSeconds"

  private val context: Context
    get() = overrideContext ?: NitroModules.applicationContext ?: throw Error("NitroModules.applicationContext not set")

  private fun prefs(ctx: Context): SharedPreferences =
    ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

  private val prefs: SharedPreferences get() = prefs(context)

  fun getStoredVersion(ctx: Context): String? = prefs(ctx).getString(KEY_VERSION, null)
  fun getStoredVersion(): String? = prefs.getString(KEY_VERSION, null)

  fun setStoredVersion(value: String?) {
    prefs.edit().apply {
      if (value != null) putString(KEY_VERSION, value) else remove(KEY_VERSION)
      apply()
    }
  }

  fun getStoredBundlePath(ctx: Context): String? = prefs(ctx).getString(KEY_BUNDLE_PATH, null)
  fun getStoredBundlePath(): String? = prefs.getString(KEY_BUNDLE_PATH, null)

  fun setStoredBundlePath(value: String?) {
    prefs.edit().apply {
      if (value != null) putString(KEY_BUNDLE_PATH, value) else remove(KEY_BUNDLE_PATH)
      apply()
    }
  }

  var isPendingValidation: Boolean
    get() = prefs.getBoolean(KEY_PENDING_VALIDATION, false)
    set(value) = prefs.edit().putBoolean(KEY_PENDING_VALIDATION, value).apply()

  var lastCheckedRemoteVersion: String?
    get() = prefs.getString(KEY_LAST_CHECKED_REMOTE_VERSION, null)
    set(value) {
      prefs.edit().apply {
        if (value != null) putString(KEY_LAST_CHECKED_REMOTE_VERSION, value) else remove(KEY_LAST_CHECKED_REMOTE_VERSION)
        apply()
      }
    }

  fun getBlacklist(): List<String> {
    val json = prefs.getString(KEY_BLACKLIST, "[]") ?: "[]"
    val arr = JSONArray(json)
    return (0 until arr.length()).map { arr.getString(it) }
  }

  fun setBlacklist(list: List<String>) {
    val arr = JSONArray(list)
    prefs.edit().putString(KEY_BLACKLIST, arr.toString()).apply()
  }

  fun getRollbackHistory(): List<Map<String, String>> {
    val json = prefs.getString(KEY_ROLLBACK_HISTORY, "[]") ?: "[]"
    val arr = JSONArray(json)
    return (0 until arr.length()).map { i ->
      val obj = arr.getJSONObject(i)
      mapOf(
        "timestamp" to obj.optString("timestamp", ""),
        "fromVersion" to obj.optString("fromVersion", ""),
        "toVersion" to obj.optString("toVersion", ""),
        "reason" to obj.optString("reason", "")
      )
    }
  }

  fun appendRollbackRecord(fromVersion: String, toVersion: String, reason: String) {
    val history = getRollbackHistory().toMutableList()
    history.add(
      mapOf(
        "timestamp" to System.currentTimeMillis().toString(),
        "fromVersion" to fromVersion,
        "toVersion" to toVersion,
        "reason" to reason
      )
    )
    val arr = org.json.JSONArray(history.map { m ->
      org.json.JSONObject(m)
    })
    prefs.edit().putString(KEY_ROLLBACK_HISTORY, arr.toString()).apply()
  }

  fun savePreviousForRollback() {
    prefs.edit()
      .putString(KEY_PREVIOUS_VERSION, getStoredVersion())
      .putString(KEY_PREVIOUS_BUNDLE_PATH, getStoredBundlePath())
      .apply()
  }

  fun getPreviousVersion(): String? = prefs.getString(KEY_PREVIOUS_VERSION, null)
  fun getPreviousBundlePath(): String? = prefs.getString(KEY_PREVIOUS_BUNDLE_PATH, null)

  fun clearPrevious() {
    prefs.edit().remove(KEY_PREVIOUS_VERSION).remove(KEY_PREVIOUS_BUNDLE_PATH).apply()
  }

  fun getBundlesDir(): File {
    val dir = File(context.applicationInfo.dataDir, "nitroupdate_bundles")
    if (!dir.exists()) dir.mkdirs()
    return dir
  }

  fun getBgVersionCheckUrl(): String? = prefs.getString(KEY_BG_VERSION_CHECK_URL, null)
  fun setBgVersionCheckUrl(value: String?) {
    prefs.edit().apply { if (value != null) putString(KEY_BG_VERSION_CHECK_URL, value) else remove(KEY_BG_VERSION_CHECK_URL); apply() }
  }
  fun getBgDownloadUrl(): String? = prefs.getString(KEY_BG_DOWNLOAD_URL, null)
  fun setBgDownloadUrl(value: String?) {
    prefs.edit().apply { if (value != null) putString(KEY_BG_DOWNLOAD_URL, value) else remove(KEY_BG_DOWNLOAD_URL); apply() }
  }
  fun getBgBundlePathInZip(): String? = prefs.getString(KEY_BG_BUNDLE_PATH_IN_ZIP, null)
  fun setBgBundlePathInZip(value: String?) {
    prefs.edit().apply { if (value != null) putString(KEY_BG_BUNDLE_PATH_IN_ZIP, value) else remove(KEY_BG_BUNDLE_PATH_IN_ZIP); apply() }
  }
  var bgIntervalSeconds: Long
    get() = prefs.getLong(KEY_BG_INTERVAL_SECONDS, 3600L).takeIf { it > 0 } ?: 3600L
    set(value) = prefs.edit().putLong(KEY_BG_INTERVAL_SECONDS, value).apply()
}

File: packages/react-native-nitro-update/ios/OtaStorage.swift
import Foundation

/// Keys and paths for OTA bundle storage (used by HybridBundleUpdater and NitroUpdateBundleLoader).
enum OtaStorage {
  private static let defaults = UserDefaults.standard
  private static let suite = "com.nitroupdate"
  
  static var userDefaults: UserDefaults {
    defaults
  }
  
  static let versionKey = "nitroupdate.version"
  static let bundlePathKey = "nitroupdate.bundlePath"
  static let pendingValidationKey = "nitroupdate.pendingValidation"
  static let blacklistKey = "nitroupdate.blacklist"
  static let rollbackHistoryKey = "nitroupdate.rollbackHistory"
  static let previousVersionKey = "nitroupdate.previousVersion"
  static let previousBundlePathKey = "nitroupdate.previousBundlePath"
  static let lastCheckedRemoteVersionKey = "nitroupdate.lastCheckedRemoteVersion"
  static let bgVersionCheckUrlKey = "nitroupdate.bgVersionCheckUrl"
  static let bgDownloadUrlKey = "nitroupdate.bgDownloadUrl"
  static let bgBundlePathInZipKey = "nitroupdate.bgBundlePathInZip"
  static let bgIntervalSecondsKey = "nitroupdate.bgIntervalSeconds"

  static var bundlesDirectory: URL {
    let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
    let dir = appSupport.appendingPathComponent("NitroUpdate/bundles", isDirectory: true)
    try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
  }
  
  static func getStoredVersion() -> String? {
    defaults.string(forKey: versionKey)
  }
  
  static func setStoredVersion(_ value: String?) {
    if let v = value { defaults.set(v, forKey: versionKey) }
    else { defaults.removeObject(forKey: versionKey) }
  }
  
  static func getStoredBundlePath() -> String? {
    defaults.string(forKey: bundlePathKey)
  }
  
  static func setStoredBundlePath(_ value: String?) {
    if let v = value { defaults.set(v, forKey: bundlePathKey) }
    else { defaults.removeObject(forKey: bundlePathKey) }
  }
  
  static var isPendingValidation: Bool {
    get { defaults.bool(forKey: pendingValidationKey) }
    set { defaults.set(newValue, forKey: pendingValidationKey) }
  }
  
  static var lastCheckedRemoteVersion: String? {
    get { defaults.string(forKey: lastCheckedRemoteVersionKey) }
    set {
      if let v = newValue { defaults.set(v, forKey: lastCheckedRemoteVersionKey) }
      else { defaults.removeObject(forKey: lastCheckedRemoteVersionKey) }
    }
  }
  
  static func getBlacklist() -> [String] {
    guard let data = defaults.data(forKey: blacklistKey),
          let arr = try? JSONDecoder().decode([String].self, from: data) else { return [] }
    return arr
  }
  
  static func setBlacklist(_ list: [String]) {
    guard let data = try? JSONEncoder().encode(list) else { return }
    defaults.set(data, forKey: blacklistKey)
  }
  
  static func getRollbackHistory() -> [[String: String]] {
    guard let data = defaults.data(forKey: rollbackHistoryKey),
          let arr = try? JSONDecoder().decode([[String: String]].self, from: data) else { return [] }
    return arr
  }
  
  static func appendRollbackRecord(fromVersion: String, toVersion: String, reason: String) {
    var history = getRollbackHistory()
    history.append([
      "timestamp": String(Int(Date().timeIntervalSince1970 * 1000)),
      "fromVersion": fromVersion,
      "toVersion": toVersion,
      "reason": reason,
    ])
    if let data = try? JSONEncoder().encode(history) {
      defaults.set(data, forKey: rollbackHistoryKey)
    }
  }
  
  static func savePreviousForRollback() {
    if let v = getStoredVersion() { defaults.set(v, forKey: previousVersionKey) }
    if let p = getStoredBundlePath() { defaults.set(p, forKey: previousBundlePathKey) }
  }
  
  static func getPreviousVersion() -> String? { defaults.string(forKey: previousVersionKey) }
  static func getPreviousBundlePath() -> String? { defaults.string(forKey: previousBundlePathKey) }

  static func getBgVersionCheckUrl() -> String? { defaults.string(forKey: bgVersionCheckUrlKey) }
  static func setBgVersionCheckUrl(_ value: String?) {
    if let v = value { defaults.set(v, forKey: bgVersionCheckUrlKey) } else { defaults.removeObject(forKey: bgVersionCheckUrlKey) }
  }
  static func getBgDownloadUrl() -> String? { defaults.string(forKey: bgDownloadUrlKey) }
  static func setBgDownloadUrl(_ value: String?) {
    if let v = value { defaults.set(v, forKey: bgDownloadUrlKey) } else { defaults.removeObject(forKey: bgDownloadUrlKey) }
  }
  static func getBgBundlePathInZip() -> String? { defaults.string(forKey: bgBundlePathInZipKey) }
  static func setBgBundlePathInZip(_ value: String?) {
    if let v = value { defaults.set(v, forKey: bgBundlePathInZipKey) } else { defaults.removeObject(forKey: bgBundlePathInZipKey) }
  }
  static var bgIntervalSeconds: TimeInterval {
    get { defaults.double(forKey: bgIntervalSecondsKey) > 0 ? defaults.double(forKey: bgIntervalSecondsKey) : 3600 }
    set { defaults.set(newValue, forKey: bgIntervalSecondsKey) }
  }

  static func clearPrevious() {
    defaults.removeObject(forKey: previousVersionKey)
    defaults.removeObject(forKey: previousBundlePathKey)
  }
}
